server/src-lib/Hasura/GraphQL/Schema.hs

```diff
@@ -1476,14 +1530,85 @@ mkGCtxMapTable tableCache (TableInfo tn _ fields rolePerms constraints pkeyCols
 noFilter :: S.BoolExp
 noFilter = S.BELit True
 
-mkScalarTyInfo :: PGColType -> ScalarTyInfo
-mkScalarTyInfo = ScalarTyInfo Nothing
+checkConflictingNodes
+  :: (MonadError QErr m)
+  => GCtx -> RemoteGCtx -> m ()
+checkConflictingNodes gCtx remoteCtx = do
+  -- check type conflicts
+  let typeMap = _gTypes gCtx
+      types = map G.unNamedType $ Map.keys typeMap
+      rmTypes = filter (`notElem` builtinTy) $
+                map G.unNamedType $ Map.keys $ _rgTypes remoteCtx
+      conflictedTypes = filter (`elem` types) rmTypes
+
+  unless (null conflictedTypes) $
+    throw400 RemoteSchemaConflicts $ tyMsg conflictedTypes
+
+  -- check node conflicts
+  let rmQRoot = _otiFields $ _rgQueryRoot remoteCtx
+      rmMRoot = _otiFields <$> _rgMutationRoot remoteCtx
+      rmRoots = filter (`notElem` builtin) . Map.keys <$> fmap (Map.union rmQRoot) rmMRoot
+      hQR = _otiFields <$>
+            join (getObjTyM <$> Map.lookup (G.NamedType "query_root") typeMap)
+      hMR = _otiFields <$>
+            join (getObjTyM <$> Map.lookup (G.NamedType "mutation_root") typeMap)
+      hRoots = Map.keys <$> mergeMaybeMaps hQR hMR
+  case (rmRoots, hRoots) of
+    (Just rmR, Just hR) -> do
+      let conflictedNodes = filter (`elem` hR) rmR
+      unless (null conflictedNodes) $
+        throw400 RemoteSchemaConflicts $ nodesMsg conflictedNodes
+    _ -> return ()
+
+  where
+    tyMsg ty = "types: [" <> nodesTxt ty <> "] already exist in current graphql schema"
+    nodesMsg n = "nodes : [" <> nodesTxt n <> "] already exist in current graphql schema"
+    nodesTxt nodes = T.intercalate ", " $ map G.unName nodes
+    builtin = ["__type", "__schema", "__typename", "Query", "Mutation", "query_root", "mutation_root"]
+    builtinTy = [ "__Directive"
```

Comment:
remove hardcoded `Query`, `Mutation` ... and filter query and mutation root types of both hasura and remote schema